maize/building-and-booting.rst

```diff
@@ -0,0 +1,186 @@
+
+While it is possible to edit the ``.config`` file by hand, the Linux kernel comes with several tools to make it more convenient to edit the configuration.
+The one we'll be using is ``menuconfig``, which has a menu-driven interface for configuring the kernel.
+
+.. code:: sh
+
+    # This will compile menuconfig if necessary, then run it.
+    myx500@cs-u-maize.cs.umn.edu:~/linux$ make menuconfig
+
+After running the above command, your terminal should be displaying a menu with a blue background.
+There are two configuration options we want to enable.
+
+The first configuration option to enable is ``CONFIG_STAGING``.
+This enables building drivers that are marked as not yet being ready for release in the main Linux kernel, but are still in the repo for users willing to try them out.
+Most new drivers enter the kernel as staging drivers, so we've marked our ``hello_world`` module as one.
+
+Use the arrow keys to navigate to the ``Device Drivers`` option, then hit :kbd:`Enter` to enter that submenu.
+Then, scroll until you find the ``Staging drivers`` option, then hit :kbd:`Space` to enable it.
+
+The second option we want to enable is ``CONFIG_HELLO_WORLD``, which enables our driver.
+
+With the cursor still on ``Staging drivers``, hit :kbd:`Enter` to enter the submenu for staging drivers.
+Then, scroll to the ``'Hello, World!' Driver`` option and hit :kbd:`Space`.
+This should make the field next to it go from ``< >`` (indicating that the module will not be built) to ``<M>`` (indicating that the module will be built, but kept as a ``.ko`` file separately from the rest of the kernel).
+Hit :kbd:`Space` again to make it go from ``<M>`` to ``<*>``, which indicates that the module will be built and linked into the rest of the kernel.
+
+.. code::
+
+    Device Drivers --->
+      [*] Staging drivers --->
+        <*> 'Hello, World!' Driver
+
+Hit :kbd:`Esc` twice to go up a menu level to the ``Staging drivers`` menu, then four more times to get to the main menu.
+
+Hit :kbd:`Esc` twice more to exit.
+``menuconfig`` will prompt you, asking if you want to save your changes.
+Press :kbd:`Enter` to save them.
+
+We've now configured our kernel.
+We can use virtme-ng again to build it.
+Notice how this time, the build is much faster, since we can re-use almost all of the output files from the previous build.
+You should also be able to see a line from compiling the ``hello_world`` driver.
+
+.. code:: sh
+
+    myx500@cs-u-maize.cs.umn.edu:~/linux$ vng --verbose --build
+    [...]
+      CC      drivers/staging/hello_world/hello_world.o
+    [...]
+
+You can use virtme-ng again to boot the kernel, too.
+
+.. code:: sh
+
+    myx500@cs-u-maize.cs.umn.edu:~/linux$ vng --verbose
+
+Our driver just prints ``Hello, World!`` as the kernel boots.
+We can see the kernel logs with the ``dmesg`` command, and pipe them through the ``grep`` command to search for ones that include "Hello".
+
+.. code:: sh
+
+    myx500@virtme-ng:~/linux$ dmesg | grep Hello
```

Comment:
Any particular reason for recommending `dmesg` over `journalctl` as was done in the earlier tutorial versions?

Edit: Oh I see, `journalctl` doesn't work, I guess because there's no systemd in the virtme-ng environment?

Reply:
Yeah, virtme-ng uses its own init script; if it used the host systemd, I think it'd be a lot easier to get into trouble with the overlayfs+9pfs stuff, since some boot services will need to access files etc. that aren't necessarily readable by the user running the VM.

(I'm also in the "never actually learned journalctl flags beyond `-f`" camp :sweat_smile:)